replica-node/src/accountsdb_repl_service.rs

```diff
@@ -0,0 +1,81 @@
+/// Module responsible for replicating AccountsDb data from its peer to its local AccountsDb
+use {
+    log::*,
+    solana_replica_lib::accountsdb_repl_client::{
+        AccountsDbReplClientService, AccountsDbReplClientServiceConfig, ReplicaRpcError,
+    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::{
+        thread::{self, sleep, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+pub struct AccountsDbReplService {
+    thread: JoinHandle<()>,
+}
+
+impl AccountsDbReplService {
+    pub fn new(
+        last_replicated_slot: Slot,
+        config: AccountsDbReplClientServiceConfig,
+    ) -> Result<Self, ReplicaRpcError> {
+        let accountsdb_repl_client = AccountsDbReplClientService::new(config)?;
+        let thread = Builder::new()
+            .name("sol-accountsdb-repl-svc".to_string())
+            .spawn(move || {
+                Self::run_service(last_replicated_slot, accountsdb_repl_client);
+            })
+            .unwrap();
+        Ok(Self { thread })
+    }
+
+    fn replicate_accounts_for_slot(
+        accountsdb_repl_client: &mut AccountsDbReplClientService,
+        slot: Slot,
+    ) {
+        match accountsdb_repl_client.get_slot_accounts(slot) {
+            Err(err) => {
+                error!(
+                    "Ran into error getting accounts for slot {:?}, error: {:?}",
+                    slot, err
+                );
+            }
+            Ok(accounts) => {
+                for account in accounts.iter() {
+                    debug!(
+                        "Received account: {:?}",
+                        Pubkey::new(&account.account_meta.as_ref().unwrap().pubkey)
+                    );
+                }
+            }
+        }
+    }
+
+    fn run_service(
+        mut last_replicated_slot: Slot,
+        mut accountsdb_repl_client: AccountsDbReplClientService,
+    ) {
+        loop {
+            match accountsdb_repl_client.get_updated_slots(last_replicated_slot) {
+                Ok(slots) => {
+                    info!("Received updated slots: {:?}", slots);
+                    if !slots.is_empty() {
+                        for slot in slots.iter() {
+                            Self::replicate_accounts_for_slot(&mut accountsdb_repl_client, *slot);
+                        }
```

Comment:
A bit worried here that before  we've fetched all the accounts, the root will have moved past the confirmed slots we just fetched, and we may have cleaned some necessary information.

For example:

`1 - > 2 -> 3 -> 4 -> 5`

Let's say the slots we returned from `get_updated_slots()` are `1` and `2`.

While we are running this loop `for slot in slots.iter() {`, the validator makes a root at 
slot `3` and cleans up information from slots `1` and `2`. Now we have an inconsistent view of slot `2`.

It seems like when we make a request for the accounts of `requested_slot` we need to fetch *all* information up to and including `safe_slot = max(requested_slot, current_root)`

Reply:
Agreed with the concern of race condition. I was planning to address it in a following-on PR. We have to address gaps, using your example,
1 - > 2 -> 3 -> 4 -> 5

Suppose we successfully replicated information for slot 1 and 2. Bank for slot 3 is dropped. And we process at 4. Will we not have consistency issue as well?

Reply:
> Suppose we successfully replicated information for slot 1 and 2. Bank for slot 3 is dropped at the server. And we process at 4. Will we not have consistency issue as well?

The server should return all the parents of the `safe_slot`, so we will know if there are any gaps missing.

Reply:
This issue will be handled in installment 3.